### PR TITLE
Fix inInference state corruption.

### LINKED ISSUE #22290: Inference results lost after multiple codegen invocations
#21677 seems to have caused a regression in irgen idempotence (verified by bisect):

```
julia> exec_dummy() = return nothing
exec_dummy (generic function with 1 method)

julia> @code_llvm exec_dummy()

; Function Attrs: sspstrong
define void @julia_exec_dummy_66504() #0 !dbg !5 {
top:
  ret void
}

julia> @code_llvm exec_dummy()

; Function Attrs: sspstrong
define void @julia_exec_dummy_66528() #0 !dbg !5 {
top:
  ret void
}

julia> # third time's a charm
julia> @code_llvm exec_dummy()

; Function Attrs: sspstrong
define i8** @japi1_exec_dummy_66528(i8**, i8***, i32) #0 !dbg !5 {
top:
  %3 = alloca i8***, align 8
  store volatile i8*** %1, i8**** %3, align 8
  ret i8** inttoptr (i64 140093546754056 to i8**)
}
```

The `japi1` emission is caused because upon third emission `specsig` is false due to `src->inferred` being false too. This breaks CUDAnative, where we can only handle specialized `julia_*` signatures. (Note we use `_dump_function` with `CodegenParams(cached=false)` over there.)

Sorry for not doing much debugging here myself, but I'm short on time.
cc @vtjnash  @jrevels

### PR DESCRIPTION
**Pull request by maleadt**
This led to non-idempotent irgen.
Fixes #22290, caused by #21677.

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

File: test/inference.jl [hunk 1]
```diff
@@ -855,4 +862,4 @@ let f, m
         Expr(:return, Expr(:call, Core._apply, :+, SSAValue(0)))
     ]
     @test @inferred(f()) == 6
-end
+end
```
Comment (ararslan) [inline test/inference.jl]: My bad, introduced when fixing conflicts

Comment (vtjnash) [inline test/inference.jl]: there's a fix needed to one of the staged tests too (to pass CI)

Comment (ararslan) [inline test/inference.jl]: Not sure what you mean. Is the fix needed due to an issue with my conflict resolution?

Comment (vtjnash) [inline test/inference.jl]: no, just a change to another number to make the CI test pass

**Comment by maleadt** [comment]
Rebased and updated failing staged function test.

File: test/staged.jl [hunk 2]
```diff
@@ -174,7 +174,7 @@ let gf_err, tsk = @async nothing # create a Task for yield to try to run
     end
     @test_throws ErrorException gf_err()
     @test_throws ErrorException gf_err()
-    @test gf_err_ref[] == 3
+    @test gf_err_ref[] == 4
```
Comment (tkelman) [inline test/staged.jl]: why does this change?

Comment (vtjnash) [inline test/staged.jl]: The value is unimportant, it just tells how many times we attempted to and failed to run the generated function and is useful for being aware of when it changes

Comment (tkelman) [inline test/staged.jl]: why does the number of times this attempted to and failed to run the generated function change then?

Comment (vtjnash) [inline test/staged.jl]: We used to disable attempts to reinfer a function if it failed. This PR changes it to always attempt to reinfer a function.

**Pull request merged**